src/main/java/org/mastodon/mamut/fitting/edgel/SampleEllipsoidEdgel.java

```diff
@@ -60,91 +64,119 @@ public static Map< Edgel, Double > getCosts(
-		final ArrayList< Integer > indices = new ArrayList<>();
+		final TIntArrayList indices = new TIntArrayList();
 		final double[][] coordinates = new double[ numPointsPerSample ][ 3 ];
 
 		Ellipsoid bestEllipsoid = null;
 		double bestCost = Double.POSITIVE_INFINITY;
 		final double[] center = new double[ 3 ];
+		int candidates = 0;
 
 		for ( int sample = 0; sample < numSamples; ++sample )
 		{
-			try
-			{
-				indices.clear();
-				for ( int s = 0; s < numPointsPerSample; ++s )
-				{
-					int i = rand.nextInt( edgels.size() );
-					while ( indices.contains( i ) )
-						i = rand.nextInt( edgels.size() );
-					indices.add( i );
-					edgels.get( i ).localize( coordinates[ s ] );
-				}
-				final Ellipsoid ellipsoid = FitEllipsoid.yuryPetrov( coordinates );
-
-				// don't count degenerate samples
-				final double[] radii = ellipsoid.getRadii();
-				if ( Double.isNaN( radii[ 0 ] ) || Double.isNaN( radii[ 1 ] ) || Double.isNaN( radii[ 2 ] ) )
-				{
-					--sample;
-					continue;
-				}
-
-				ellipsoid.localize( center );
-				LinAlgHelpers.subtract( center, expectedCenter, center );
-				if ( LinAlgHelpers.length( center ) > maxCenterDistance )
-					continue;
-
-				final double cost = costFunction.compute( ellipsoid, edgels );
-				if ( cost < bestCost )
-				{
-					bestCost = cost;
-					bestEllipsoid = ellipsoid;
-				}
-			}
-			catch ( final IllegalArgumentException e )
-			{
-//				e.printStackTrace();
-//				System.out.println( "oops" );
-			}
-			catch ( final RuntimeException e )
+			sampleCoordinatesFromEdgels( edgels, coordinates, rand, indices );
+
+			final Ellipsoid ellipsoid = tryFitEllipsoidYuryPetrov( coordinates );
+
+			if ( ! isEllipsoidValid( ellipsoid, expectedCenter, maxCenterDistance, center ) )
+				continue;
+
+			final double cost = costFunction.compute( ellipsoid, edgels );
+			if ( cost < bestCost )
 			{
-//				System.out.println( "psd" );
+				bestCost = cost;
+				bestEllipsoid = ellipsoid;
 			}
+
+			if( ++candidates >= numCandidates )
+				break;
 		}
 
-		if ( bestEllipsoid == null )
-			return null;
+		if ( bestEllipsoid == null ) // no ellipsoid found
+			throw new NoEllipsoidFoundException( "No ellipsoid found, that is near to the expected center." );
 
-		final Ellipsoid refined = fitToInliers( bestEllipsoid, edgels, outsideCutoffDistance, insideCutoffDistance, angleCutoffDistance );
-		if ( refined == null )
+		try
+		{
+			// refined ellipsoid
+			return fitToInliers( bestEllipsoid, edgels, outsideCutoffDistance, insideCutoffDistance,
+					angleCutoffDistance );
+		}
+		catch ( final RuntimeException e )
 		{
-//			System.err.println( "refined ellipsoid == null! This shouldn't happen!");
+			// refinement failed
 			return bestEllipsoid;
 		}
-		return refined;
+	}
+
+	private static void sampleCoordinatesFromEdgels( List< Edgel > edgels, double[][] coordinates, Random rand, TIntArrayList indices )
+	{
+		indices.clear();
+		for ( int s = 0; s < coordinates.length; ++s )
```

Comment:
```suggestion
		for ( int s = 0; s < coordinates.length; s++ )
```
Pre- and Post-increment in the loop definition does not make a difference here - as far as I understand. I would prefer post-increment, since this is the more usual way in Java. Otherwise, one has to think, why it is different here.

Reply:
Let's fix these pre increments in one commit for the whole repo in a separate PR.

Reply:
I create a new issue #6. Please resolve.